coalib/misc/Caching.py

```diff
@@ -0,0 +1,168 @@
+import appdirs
+import calendar
+import os
+import pickle
+import time
+
+from coalib.misc.Constants import caching_db, changed_files_db
+
+
+def get_config_data_path(filename):
+    """
+    Get the full path of ``filename`` present in the user config directory.
+
+    :param filename: The file whose path needs to be expanded.
+    :return:         Full path of the file, assuming it's present in the
+                     user's config directory.
+    """
+    path = appdirs.user_config_dir("coala")
+    return os.path.join(path, filename)
+
+
+def pickle_load(filename, fallback=None):
+    """
+    Get the data stored in ``filename`` present in the coala user
+    config directory.
+
+    :param filename: The name of the file present in the coala user config
+                     directory.
+    :param fallback: Return value to fallback to in case the file doesn't
+                     exist.
+    :return:         Data that is present in the file, if the file exists.
+                     Otherwise the ``default`` value is returned.
+    """
+    filename = get_config_data_path(filename)
+    if not os.path.isfile(filename):
+        return fallback
+    with open(filename, "rb") as f:
+        return pickle.load(f)
+
+
+def pickle_dump(filename, data):
+    """
+    Write ``data`` into the file ``filename`` present in the coala user
+    config directory.
+
+    :param filename: The name of the file present in the coala user config
+                     directory.
+    :param data:     Data to be serialized and written to the file using
+                     pickle.
+    """
+    filename = get_config_data_path(filename)
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+    with open(filename, "wb") as f:
+        return pickle.dump(data, f)
+
+
+def verify_file_integrity(log_printer, files=[caching_db, changed_files_db]):
+    """
+    Verify that the pickled database files are not corrupted.
```

Comment:
so this essentially loads and throws away all cached data which is bad for performance since it's IO and IO is always slow, shouldn't the pickle error get thrown on load also normally? Can't we handle it there?